onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Shukant Pal.
+// Licensed under the MIT License.
+
+#include "core/providers/common.h"
+#include "core/providers/shared/utils/utils.h"
+
+#ifdef __APPLE__
+#include "core/providers/coreml/builders/model_builder.h"
+#endif
+#include "core/providers/coreml/builders/helper.h"
+#include "core/providers/coreml/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class ReductionOpBuilder : public BaseOpBuilder {
+ private:
+#ifdef __APPLE__
+  [[nodiscard]] Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                             const logging::Logger& logger) const override;
+#endif
+};
+
+namespace {
+template <typename T>
+void AddReductionParams(T* params, std::vector<int64_t>& axes, bool keepdims) {
+  params->set_keepdims(keepdims);
+  for (auto& axis : axes) {
+    params->add_axes(axis);
+  }
+}
+} // namespace
+
+#ifdef __APPLE__
+Status ReductionOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                                 const logging::Logger& /* logger */) const {
+  const auto& op_type(node.OpType());
+  const auto& input_defs(node.InputDefs());
+
+  NodeAttrHelper helper(node);
+  auto axes = helper.Get("axes", std::vector<int64_t>{});
```

Comment:
in ONNX opset 18, 'axes' is an input instead of an attribute. this changed earlier in opset 13 for ReduceSum.
could possibly first check for the presence of the second, optional, 'axes' input and ensure that it is constant (a constant initializer). if it doesn't exist, then fall back to checking for the attribute.
that probably should be in an override of IsOpSupported(), as a non-constant axes input would not be supported.